pretty: render tabs at the end of the pipeline

knz opened the pull request:
Prior to this patch, tabs were expanded into a string prefix early,
during the construction of the document. This caused overhead during
the selection of the best layout, as the prefixes needed to be
re-composed for each layout alternative.

This patch instead defers the expansion of tabs to the very end,
during the `layout` phase.

```
name           old time/op    new time/op    delta
PrettyData-16     200ms ± 9%     168ms ± 9%  -16.04%  (p=0.008 n=5+5)

name           old alloc/op   new alloc/op   delta
PrettyData-16    26.8MB ± 0%    22.2MB ± 0%  -17.21%  (p=0.008 n=5+5)

name           old allocs/op  new allocs/op  delta
PrettyData-16     12.1k ± 0%      6.6k ± 0%  -45.32%  (p=0.008 n=5+5)
```

Release note: None

knz committed:
pretty: render tabs at the end of the pipeline

Prior to this patch, tabs were expanded into a string prefix early,
during the construction of the document. This caused overhead during
the selection of the best layout, as the prefixes needed to be
re-composed for each layout alternative.

This patch instead defers the expansion of tabs to the very end,
during the `layout` phase.

```
name           old time/op    new time/op    delta
PrettyData-16     200ms ± 9%     168ms ± 9%  -16.04%  (p=0.008 n=5+5)

name           old alloc/op   new alloc/op   delta
PrettyData-16    26.8MB ± 0%    22.2MB ± 0%  -17.21%  (p=0.008 n=5+5)

name           old allocs/op  new allocs/op  delta
PrettyData-16     12.1k ± 0%      6.6k ± 0%  -45.32%  (p=0.008 n=5+5)
```

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/27259)
<!-- Reviewable:end -->

knz committed:
pretty: add a missing comment

Release note: None

knz committed:
pretty: rename "nestname" to "nestunder"

The header of a "nested section" can be anything, not just a name. So
make the function name more generic.

Release note: None

madelynnblue reviewed (approved):
Ohhh pretty great.

knz commented:
Thank you :-)
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

knz commented:
bors r+

mjibson commented on pkg/util/pretty/util.go:
```diff
@@ -133,6 +133,14 @@ func FoldMap(f func(a, b Doc) Doc, g func(Doc) Doc, d ...Doc) Doc {
 
 // Bracket brackets x with l and r and given Nest arguments.
 func Bracket(n int, l string, x Doc, r string) Doc {
+	// The "straightforward" implementation of Bracket should really be:
```
This comment isn't accurate. The reason we can't use Group is because we need `Line` to be the empty string if it is flattened, instead of the normal `" "`.

knz replied:
Hugh, this was totally invisible to me. It would have helped if that was documented :)

I'll add the correct comment.

knz replied:
Done:  #27263
